cql_tracing_test.TestCqlTracing.tracing_startup_test hangs on shutdown in dtest debug

Regression started at https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-debug/721/artifact/logs-debug.2/1612888868753_cql_tracing_test.TestCqlTracing.tracing_startup_test/node2.log
```
INFO  2021-02-09 16:26:10,988 [shard 0] compaction_manager - Asked to stop
INFO  2021-02-09 16:26:10,989 [shard 0] compaction_manager - Stopping 0 ongoing compactions
INFO  2021-02-09 16:26:10,989 [shard 0] init - Signal received; shutting down
INFO  2021-02-09 16:26:10,989 [shard 0] init - Shutting down redis service
INFO  2021-02-09 16:26:10,989 [shard 0] init - Shutting down redis service was successful
INFO  2021-02-09 16:26:10,990 [shard 0] init - Shutting down view builder
INFO  2021-02-09 16:26:10,990 [shard 0] view - Stopping view builder
INFO  2021-02-09 16:26:11,007 [shard 1] compaction_manager - Asked to stop
INFO  2021-02-09 16:26:11,008 [shard 1] compaction_manager - Stopping 0 ongoing compactions
INFO  2021-02-09 16:26:11,028 [shard 0] compaction_manager - Stopped
INFO  2021-02-09 16:26:11,030 [shard 1] view - Stopping view builder
INFO  2021-02-09 16:26:11,087 [shard 1] compaction_manager - Stopped
INFO  2021-02-09 16:26:11,210 [shard 0] storage_service - Node 127.0.12.1 state jump to normal
INFO  2021-02-09 16:26:11,326 [shard 0] init - Shutting down view builder was successful
INFO  2021-02-09 16:26:11,326 [shard 0] init - Shutting down local storage
INFO  2021-02-09 16:26:11,329 [shard 0] storage_service - Drain on shutdown: starts
INFO  2021-02-09 16:26:11,331 [shard 0] storage_service - Stop transport: starts
INFO  2021-02-09 16:26:11,332 [shard 0] storage_service - Shutting down native transport
INFO  2021-02-09 16:26:11,817 [shard 0] cql_server_controller - CQL server stopped
INFO  2021-02-09 16:26:11,832 [shard 0] storage_service - Shutting down native transport was successful
INFO  2021-02-09 16:26:11,832 [shard 0] storage_service - Shutting down rpc server
INFO  2021-02-09 16:26:11,884 [shard 0] thrift_controller - Thrift server stopped
INFO  2021-02-09 16:26:11,936 [shard 0] storage_service - Shutting down rpc server was successful
INFO  2021-02-09 16:26:11,936 [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
INFO  2021-02-09 16:26:11,938 [shard 0] gossip - My status = NORMAL
INFO  2021-02-09 16:26:11,938 [shard 0] gossip - Announcing shutdown
INFO  2021-02-09 16:26:12,027 [shard 0] storage_service - Node 127.0.12.2 state jump to normal
INFO  2021-02-09 16:26:14,386 [shard 0] gossip - Disable and wait for gossip loop started
INFO  2021-02-09 16:26:14,386 [shard 0] gossip - Disable and wait for gossip loop finished
INFO  2021-02-09 16:26:14,400 [shard 0] gossip - Gossip is now stopped
INFO  2021-02-09 16:26:14,401 [shard 0] storage_service - Stop transport: stop_gossiping done
INFO  2021-02-09 16:26:14,462 [shard 0] storage_service - messaging_service stopped
INFO  2021-02-09 16:26:14,462 [shard 0] storage_service - Stop transport: shutdown messaging_service done
INFO  2021-02-09 16:26:14,463 [shard 0] storage_service - stream_manager stopped
INFO  2021-02-09 16:26:14,463 [shard 0] storage_service - Stop transport: shutdown stream_manager done
INFO  2021-02-09 16:26:14,463 [shard 0] storage_service - Stop transport: done
INFO  2021-02-09 16:26:14,463 [shard 0] storage_service - Drain on shutdown: stop_transport done
INFO  2021-02-09 16:26:14,464 [shard 0] tracing - Asked to shut down
INFO  2021-02-09 16:26:14,470 [shard 1] tracing - Asked to shut down
INFO  2021-02-09 16:26:14,475 [shard 0] tracing - Tracing is down
INFO  2021-02-09 16:26:15,465 [shard 1] tracing - Tracing is down
INFO  2021-02-09 16:26:15,467 [shard 0] storage_service - Drain on shutdown: tracing is stopped
INFO  2021-02-09 16:26:15,468 [shard 0] storage_service - Drain on shutdown: system distributed keyspace stopped
INFO  2021-02-09 16:26:15,470 [shard 0] hints_manager - Asked to stop
INFO  2021-02-09 16:26:15,470 [shard 0] hints_manager - Asked to stop
INFO  2021-02-09 16:26:15,470 [shard 0] hints_manager - Stopped
INFO  2021-02-09 16:26:15,470 [shard 0] hints_manager - Stopped
INFO  2021-02-09 16:26:15,472 [shard 1] hints_manager - Asked to stop
INFO  2021-02-09 16:26:15,473 [shard 1] hints_manager - Asked to stop
INFO  2021-02-09 16:26:15,475 [shard 1] hints_manager - Stopped
```

Scylla version c16e4a04239798c41fa140c14eabb250b5422983
Last passed at e1261d10f17c2d219f159c01ddf7fd2b122f9d6a
`git log --oneline e1261d10f17c2d219f159c01ddf7fd2b122f9d6a..c16e4a04239798c41fa140c14eabb250b5422983`
c16e4a042 migration_manager: Propagate schema changes with reads like we do on writes
4082f57ed Merge 'Make commitlog disk limit a hard limit.' from Calle Wilund
af2d1fa0d Update abseil submodule
b9a5aff7a distributed_loader: drop execute_futures function
a05adb853 database: Remove global storage proxy reference
8490c9ff6 transport: Remove global storage service reference
4498bb0a4 API: Fix aggregation in column_familiy
4be718ebf commitlog: Force earlier cycle/flush iff segment reserve is empty
be8c359a6 commitlog: Make segment allocation wait iff disk usage > max
ab55a1b4e commitlog: Do partial (memtable) flushing based on threshold
7c84b16cd commitlog: Make flush threshold configurable
c3d95811d table: Add a flush RP mark to table, and shortcut if not above

Today observed in another test, `test_recover_stuck_raft_upgrade`
https://jenkins.scylladb.com/job/scylla-master/job/next/5616/artifact/testlog/x86_64/dev/topology_raft_disabled.test_raft_upgrade.3.log

```
>       await asyncio.gather(*(restart(manager, srv) for srv in others))
...
>           raise asyncio.TimeoutError from None
E           TimeoutError
```

```
00:53:28.083 INFO> Using cluster: ScyllaCluster(name: 3a9acbc4-9dcc-11ed-9d58-b083fedd3c5a, running: ScyllaServer(171, 127.231.178.18, 1b4e660b-da73-4735-8477-243e605935c8), ScyllaServer(176, 127.231.178.30, 658cd4c7-7f73-4506-b4d6-1c9a6142ab71), ScyllaServer(183, 127.231.178.17, 0a39379e-b85a-4b48-8cc9-67fa8fba7714), stopped: ) for test test_recover_stuck_raft_upgrade
```

The node which hanged during shutdown was 176
https://jenkins.scylladb.com/job/scylla-master/job/next/5616/artifact/testlog/x86_64/dev/scylla-176.log
the last log lines are:
```
INFO  2023-01-27 00:54:01,635 [shard 0] hints_manager - Asked to stop
INFO  2023-01-27 00:54:01,635 [shard 0] hints_manager - Stopped
INFO  2023-01-27 00:54:01,635 [shard 0] hints_manager - Asked to stop
INFO  2023-01-27 00:54:01,635 [shard 1] hints_manager - Asked to stop
INFO  2023-01-27 00:54:01,635 [shard 1] hints_manager - Stopped
INFO  2023-01-27 00:54:01,635 [shard 1] hints_manager - Asked to stop
INFO  2023-01-27 00:54:01,635 [shard 1] hints_manager - Stopped
```

the node was killed after timeout.

Hit again apparently, in https://jenkins.scylladb.com/job/scylla-master/job/build/1423/artifact/testlog/aarch64/debug/topology_raft_disabled.test_raft_upgrade.1.log
```
______________________ test_recovery_after_majority_loss _______________________

manager = <test.pylib.manager_client.ManagerClient object at 0xffff863520d0>
random_tables = <test.pylib.random_tables.RandomTables object at 0xffff7716d750>

    @pytest.mark.asyncio
    @log_run_time
    async def test_recovery_after_majority_loss(manager: ManagerClient, random_tables: RandomTables):
        """
        We successfully upgrade a cluster. Eventually however all servers but one fail - group 0
        is left without a majority. We create a new group 0 by entering RECOVERY, using `removenode`
        to get rid of the other servers, clearing Raft data and restarting. The Raft upgrade procedure
        runs again to establish a single-node group 0. We also verify that schema changes performed
        using the old group 0 are still there.
        Note: in general there's no guarantee that all schema changes will be present; the minority
        used to recover group 0 might have missed them. However in this test the driver waits
        for schema agreement to complete before proceeding, so we know that every server learned
        about the schema changes.
    
        kbr-: the test takes about 22 seconds in dev mode on my laptop.
        """
        servers = await manager.running_servers()
    
        logging.info(f"Enabling Raft on {servers} and restarting")
        await asyncio.gather(*(enable_raft_and_restart(manager, srv) for srv in servers))
        cql = await reconnect_driver(manager)
    
        logging.info("Cluster restarted, waiting until driver reconnects to every server")
        hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
    
        logging.info(f"Driver reconnected, hosts: {hosts}. Waiting until upgrade finishes")
        await asyncio.gather(*(wait_until_upgrade_finishes(cql, h, time.time() + 60) for h in hosts))
    
        logging.info("Upgrade finished. Creating a bunch of tables")
        tables = await asyncio.gather(*(random_tables.add_table(ncolumns=5) for _ in range(5)))
    
        srv1, *others = servers
    
        logging.info(f"Killing all nodes except {srv1}")
        await asyncio.gather(*(manager.server_stop_gracefully(srv.server_id) for srv in others))
    
        logging.info(f"Entering recovery state on {srv1}")
        host1 = next(h for h in hosts if h.address == srv1.ip_addr)
        await cql.run_async("update system.scylla_local set value = 'recovery' where key = 'group0_upgrade_state'", host=host1)
>       await restart(manager, srv1)

test/topology_raft_disabled/test_raft_upgrade.py:324: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/topology_raft_disabled/test_raft_upgrade.py:37: in restart
    await manager.server_stop_gracefully(server.server_id)
test/pylib/manager_client.py:139: in server_stop_gracefully
    await self.client.get_text(f"/cluster/server/{server_id}/stop_gracefully")
test/pylib/rest_client.py:82: in get_text
    ret = await self._fetch("GET", resource_uri, response_type = "text", host = host,
test/pylib/rest_client.py:64: in _fetch
    async with request(method, uri,
/usr/lib64/python3.11/site-packages/aiohttp/client.py:1189: in __aenter__
    self._resp = await self._coro
/usr/lib64/python3.11/site-packages/aiohttp/client.py:560: in _request
    await resp.start(conn)
/usr/lib64/python3.11/site-packages/aiohttp/client_reqrep.py:894: in start
    with self._timer:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <aiohttp.helpers.TimerContext object at 0xffff77460c10>
exc_type = <class 'asyncio.exceptions.CancelledError'>
exc_val = CancelledError(), exc_tb = <traceback object at 0xffff772bb400>

    def __exit__(
        self,
        exc_type: Optional[Type[BaseException]],
        exc_val: Optional[BaseException],
        exc_tb: Optional[TracebackType],
    ) -> Optional[bool]:
        if self._tasks:
            self._tasks.pop()
    
        if exc_type is asyncio.CancelledError and self._cancelled:
>           raise asyncio.TimeoutError from None
E           TimeoutError
```

topology_raft_disabled.test_raft_upgrade.1.log:
```
07:23:06.494 INFO> Killing all nodes except Server(211, 127.17.167.39)
07:23:06.494 DEBUG> ManagerClient stopping gracefully 212
07:23:06.494 DEBUG> RESTClient fetching GET http+unix://api/cluster/server/212/stop_gracefully
07:23:06.495 DEBUG> ManagerClient stopping gracefully 213
07:23:06.495 DEBUG> RESTClient fetching GET http+unix://api/cluster/server/213/stop_gracefully
07:23:06.526 DEBUG> Message pushed from server: <EventMessage(event_type='STATUS_CHANGE', event_args={'change_type': 'DOWN', 'address': ('127.17.167.20', 9042)}, stream_id=-1, trace_id=None)>
07:23:06.547 DEBUG> Connection <AsyncoreConnection(281472683252432) 127.17.167.20:9042> closed by server
07:23:06.547 DEBUG> Closing connection (281472683252432) to 127.17.167.20:9042
07:23:06.547 DEBUG> Closed socket to 127.17.167.20:9042
07:23:14.049 INFO> Entering recovery state on Server(211, 127.17.167.39)
07:23:14.066 INFO> Stopping Server(211, 127.17.167.39) gracefully
...
07:28:14.950 DEBUG> after_test for test_recovery_after_majority_loss (success: False)
07:28:14.950 DEBUG> RESTClient fetching GET http+unix://api/cluster/after-test/False
07:28:14.952 INFO> Cluster after test test_recovery_after_majority_loss: ScyllaCluster(name: d1ed22c8-d81f-11ed-bfbc-0afa8b88e8c3, running: ScyllaServer(211, 127.17.167.39, 6def9dbc-9ab9-4261-bcb5-6ba1d8e6c04e), stopped: ScyllaServer(213, 127.17.167.11, f703b6b9-f03e-4553-9950-edc8cab37a6b), ScyllaServer(212, 127.17.167.20, bd39a7f6-233c-4752-96cc-61903b04003f))
```

https://jenkins.scylladb.com/job/scylla-master/job/build/1423/artifact/testlog/aarch64/debug/scylla-211.log shows it's hung on shitdown.  Last printout is `Shutting down auth service`
and then suppressed backtraces (reactor stalls?)

@xemul does that look familiar?

```
INFO  2023-04-11 07:23:07,227 [shard 0] gossip - Got shutdown message from 127.17.167.11, received_generation=1681186958, local_generation=1681186958
INFO  2023-04-11 07:23:07,229 [shard 0] gossip - InetAddress 127.17.167.11 is now DOWN, status = shutdown
INFO  2023-04-11 07:23:07,257 [shard 0] gossip - Got shutdown message from 127.17.167.20, received_generation=1681186955, local_generation=1681186955
INFO  2023-04-11 07:23:07,259 [shard 0] gossip - InetAddress 127.17.167.20 is now DOWN, status = shutdown
INFO  2023-04-11 07:23:10,163 [shard 0] raft_group_registry - marking Raft server f703b6b9-f03e-4553-9950-edc8cab37a6b as dead for raft groups
INFO  2023-04-11 07:23:10,255 [shard 0] raft_group_registry - marking Raft server bd39a7f6-233c-4752-96cc-61903b04003f as dead for raft groups
INFO  2023-04-11 07:23:14,070 [shard 0] compaction_manager - Asked to stop
WARN  2023-04-11 07:23:14,071 [shard 0] gossip - failure_detector_loop: Got error in the loop, live_nodes=[]: seastar::sleep_aborted (Sleep is aborted)
INFO  2023-04-11 07:23:14,072 [shard 0] gossip - failure_detector_loop: Finished main loop
INFO  2023-04-11 07:23:14,072 [shard 0] init - Signal received; shutting down
INFO  2023-04-11 07:23:14,072 [shard 0] init - Shutting down view builder ops
INFO  2023-04-11 07:23:14,072 [shard 0] view - Draining view builder
INFO  2023-04-11 07:23:14,073 [shard 1] compaction_manager - Asked to stop
INFO  2023-04-11 07:23:14,074 [shard 1] view - Draining view builder
INFO  2023-04-11 07:23:14,076 [shard 1] compaction_manager - Stopped
INFO  2023-04-11 07:23:14,082 [shard 0] compaction_manager - Stopped
INFO  2023-04-11 07:23:14,083 [shard 0] init - Shutting down view builder ops was successful
INFO  2023-04-11 07:23:14,083 [shard 0] init - Shutting down local storage
INFO  2023-04-11 07:23:14,085 [shard 0] storage_service - Stop transport: starts
INFO  2023-04-11 07:23:14,085 [shard 0] migration_manager - stopping migration service
INFO  2023-04-11 07:23:14,086 [shard 1] migration_manager - stopping migration service
INFO  2023-04-11 07:23:14,087 [shard 0] storage_service - Shutting down native transport server
INFO  2023-04-11 07:23:14,122 [shard 0] cql_server_controller - CQL server stopped
INFO  2023-04-11 07:23:14,122 [shard 0] storage_service - Shutting down native transport server was successful
INFO  2023-04-11 07:23:14,122 [shard 0] storage_service - Shutting down rpc server
INFO  2023-04-11 07:23:14,123 [shard 0] storage_service - Shutting down rpc server was successful
INFO  2023-04-11 07:23:14,123 [shard 0] storage_service - Shutting down alternator server
INFO  2023-04-11 07:23:14,126 [shard 0] storage_service - Shutting down alternator server was successful
INFO  2023-04-11 07:23:14,126 [shard 0] storage_service - Shutting down redis server
INFO  2023-04-11 07:23:14,126 [shard 0] storage_service - Shutting down redis server was successful
INFO  2023-04-11 07:23:14,127 [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
INFO  2023-04-11 07:23:14,129 [shard 0] gossip - My status = NORMAL
INFO  2023-04-11 07:23:14,129 [shard 0] gossip - Announcing shutdown
INFO  2023-04-11 07:23:14,168 [shard 0] storage_service - Node 127.17.167.39 state jump to normal
INFO  2023-04-11 07:23:17,447 [shard 0] gossip - Disable and wait for gossip loop started
INFO  2023-04-11 07:23:17,449 [shard 0] gossip - Gossip is now stopped
INFO  2023-04-11 07:23:17,450 [shard 0] storage_service - Stop transport: stop_gossiping done
INFO  2023-04-11 07:23:17,450 [shard 0] messaging_service - Stopping nontls server
INFO  2023-04-11 07:23:17,450 [shard 0] messaging_service - Stopping tls server
INFO  2023-04-11 07:23:17,451 [shard 0] messaging_service - Stopping client for address: 127.17.167.20:0
INFO  2023-04-11 07:23:17,451 [shard 0] messaging_service - Stopping client for address: 127.17.167.11:0
INFO  2023-04-11 07:23:17,451 [shard 0] messaging_service - Stopping client for address: 127.17.167.11:0
INFO  2023-04-11 07:23:17,451 [shard 0] messaging_service - Stopping client for address: 127.17.167.20:0
INFO  2023-04-11 07:23:17,452 [shard 0] messaging_service - Stopping client for address: 127.17.167.39:0
INFO  2023-04-11 07:23:17,453 [shard 0] messaging_service - Stopping client for address: 127.17.167.11:0 - Done
INFO  2023-04-11 07:23:17,453 [shard 0] messaging_service - Stopping client for address: 127.17.167.20:0 - Done
INFO  2023-04-11 07:23:17,453 [shard 0] messaging_service - Stopping client for address: 127.17.167.20:0 - Done
INFO  2023-04-11 07:23:17,453 [shard 0] messaging_service - Stopping client for address: 127.17.167.11:0 - Done
INFO  2023-04-11 07:23:17,454 [shard 0] messaging_service - Stopping tls server - Done
INFO  2023-04-11 07:23:17,454 [shard 1] messaging_service - Stopping nontls server
INFO  2023-04-11 07:23:17,454 [shard 1] messaging_service - Stopping tls server
INFO  2023-04-11 07:23:17,454 [shard 1] messaging_service - Stopping client for address: 127.17.167.11:0
INFO  2023-04-11 07:23:17,455 [shard 1] messaging_service - Stopping tls server - Done
INFO  2023-04-11 07:23:17,455 [shard 1] messaging_service - Stopping client for address: 127.17.167.11:0 - Done
INFO  2023-04-11 07:23:17,456 [shard 1] messaging_service - Stopping nontls server - Done
INFO  2023-04-11 07:23:17,456 [shard 0] messaging_service - Stopping client for address: 127.17.167.39:0 - Done
INFO  2023-04-11 07:23:17,457 [shard 0] messaging_service - Stopping nontls server - Done
INFO  2023-04-11 07:23:17,457 [shard 0] storage_service - messaging_service stopped
INFO  2023-04-11 07:23:17,457 [shard 0] storage_service - Stop transport: shutdown messaging_service done
INFO  2023-04-11 07:23:17,458 [shard 0] storage_service - Stop transport: shutdown stream_manager done
INFO  2023-04-11 07:23:17,458 [shard 0] storage_service - Stop transport: done
INFO  2023-04-11 07:23:17,460 [shard 0] tracing - Asked to shut down
INFO  2023-04-11 07:23:17,460 [shard 0] tracing - Tracing is down
INFO  2023-04-11 07:23:17,460 [shard 1] tracing - Asked to shut down
INFO  2023-04-11 07:23:17,461 [shard 1] tracing - Tracing is down
INFO  2023-04-11 07:23:17,461 [shard 0] batchlog_manager - Asked to drain
INFO  2023-04-11 07:23:17,462 [shard 0] batchlog_manager - Drained
INFO  2023-04-11 07:23:17,462 [shard 1] batchlog_manager - Asked to drain
INFO  2023-04-11 07:23:17,462 [shard 1] batchlog_manager - Drained
INFO  2023-04-11 07:23:17,463 [shard 0] compaction_manager - Asked to drain
INFO  2023-04-11 07:23:17,463 [shard 0] compaction_manager - Drained
INFO  2023-04-11 07:23:17,464 [shard 1] compaction_manager - Asked to drain
INFO  2023-04-11 07:23:17,464 [shard 1] compaction_manager - Drained
INFO  2023-04-11 07:23:17,465 [shard 0] database - Flushing non-system tables
INFO  2023-04-11 07:23:17,465 [shard 1] database - Flushing non-system tables
INFO  2023-04-11 07:23:17,480 [shard 0] database - Flushed non-system tables
INFO  2023-04-11 07:23:17,589 [shard 1] database - Flushed non-system tables
INFO  2023-04-11 07:23:17,589 [shard 0] database - Flushing system tables
INFO  2023-04-11 07:23:17,590 [shard 1] database - Flushing system tables
Reactor stalled for 33 ms on shard 1. Backtrace: 0xce47e57 0xffff8810f23b 0xffff8810e653 0xffff87f1da97 0xffff87f17de3 0xffff87f17847 0xffff87f18213 0xffff87f1f073 0x7db 0x1278eb07 0x1278e9ab 0x1278e27f 0x1278d717 0x1278d297 0x1278cf6f 0x1278cd87 0x1278cb2f 0x1276b777 0x12848263 0x128472af 0x12846b63 0x1284697b 0x12846723 0x12846497 0x1284636b 0x120bd33f 0x11d99c8b 0x120d1c5f 0x11ddbd4f 0x12c91507 0x13753a9f 0x12849b73 0x12849633 0x12849083 0x11ebe41b 0x11ebdf6f 0x11ebdd53 0x11ebdb3f 0x11ebd917 0x11ebd6eb 0x11ebd53b 0x11ebd30f 0xffff87b53c0f 0xffff8850fde3
INFO  2023-04-11 07:23:18,276 [shard 0] raft_group_registry - marking Raft server 6def9dbc-9ab9-4261-bcb5-6ba1d8e6c04e as dead for raft groups
INFO  2023-04-11 07:23:18,520 [shard 1] database - Flushed system tables
INFO  2023-04-11 07:23:20,596 [shard 0] database - Flushed system tables
INFO  2023-04-11 07:23:20,610 [shard 0] init - Shutting down local storage was successful
INFO  2023-04-11 07:23:20,610 [shard 0] init - Shutting down view update generator
INFO  2023-04-11 07:23:20,611 [shard 0] view_update_generator - leaving 0 unstaged sstables unprocessed
INFO  2023-04-11 07:23:20,611 [shard 1] view_update_generator - leaving 0 unstaged sstables unprocessed
INFO  2023-04-11 07:23:20,614 [shard 0] init - Shutting down view update generator was successful
INFO  2023-04-11 07:23:20,614 [shard 0] init - Shutting down service level controller update loop
INFO  2023-04-11 07:23:20,616 [shard 0] service_level_controller - update_from_distributed_data: configuration polling loop aborted
INFO  2023-04-11 07:23:20,617 [shard 0] init - Shutting down service level controller update loop was successful
INFO  2023-04-11 07:23:20,617 [shard 0] init - Shutting down repair
INFO  2023-04-11 07:23:20,618 [shard 0] init - Shutting down repair was successful
INFO  2023-04-11 07:23:20,618 [shard 0] init - Shutting down rpc controller API
INFO  2023-04-11 07:23:20,619 [shard 0] init - Shutting down rpc controller API was successful
INFO  2023-04-11 07:23:20,619 [shard 0] init - Shutting down transport controller API
INFO  2023-04-11 07:23:20,620 [shard 0] init - Shutting down transport controller API was successful
INFO  2023-04-11 07:23:20,620 [shard 0] init - Shutting down native transport
INFO  2023-04-11 07:23:20,621 [shard 0] init - Shutting down native transport was successful
INFO  2023-04-11 07:23:20,621 [shard 0] init - Shutting down view builder API
INFO  2023-04-11 07:23:20,622 [shard 0] init - Shutting down view builder API was successful
INFO  2023-04-11 07:23:20,622 [shard 0] init - Shutting down view builder
INFO  2023-04-11 07:23:20,622 [shard 0] view - Stopping view builder
INFO  2023-04-11 07:23:20,622 [shard 1] view - Stopping view builder
INFO  2023-04-11 07:23:20,624 [shard 0] init - Shutting down view builder was successful
INFO  2023-04-11 07:23:20,625 [shard 0] init - Shutting down hinted handoff API
INFO  2023-04-11 07:23:20,626 [shard 0] init - Shutting down hinted handoff API was successful
INFO  2023-04-11 07:23:20,626 [shard 0] init - Shutting down view update backlog broker
INFO  2023-04-11 07:23:20,628 [shard 0] init - Shutting down view update backlog broker was successful
INFO  2023-04-11 07:23:20,628 [shard 0] init - Shutting down cf cache hit rate calculator
INFO  2023-04-11 07:23:20,630 [shard 0] init - Shutting down cf cache hit rate calculator was successful
INFO  2023-04-11 07:23:20,630 [shard 0] init - Shutting down load meter
INFO  2023-04-11 07:23:20,630 [shard 0] init - Shutting down load meter was successful
INFO  2023-04-11 07:23:20,630 [shard 0] init - Shutting down batchlog manager
INFO  2023-04-11 07:23:20,631 [shard 0] batchlog_manager - Asked to stop
INFO  2023-04-11 07:23:20,631 [shard 0] batchlog_manager - Asked to drain
INFO  2023-04-11 07:23:20,631 [shard 0] batchlog_manager - Drained
INFO  2023-04-11 07:23:20,631 [shard 0] batchlog_manager - Stopped
INFO  2023-04-11 07:23:20,631 [shard 1] batchlog_manager - Asked to stop
INFO  2023-04-11 07:23:20,632 [shard 1] batchlog_manager - Asked to drain
INFO  2023-04-11 07:23:20,632 [shard 1] batchlog_manager - Drained
INFO  2023-04-11 07:23:20,632 [shard 1] batchlog_manager - Stopped
INFO  2023-04-11 07:23:20,633 [shard 0] init - Shutting down batchlog manager was successful
INFO  2023-04-11 07:23:20,634 [shard 0] init - Shutting down snapshots API
INFO  2023-04-11 07:23:20,635 [shard 0] init - Shutting down snapshots API was successful
INFO  2023-04-11 07:23:20,635 [shard 0] init - Shutting down snapshots
INFO  2023-04-11 07:23:20,636 [shard 0] init - Shutting down snapshots was successful
INFO  2023-04-11 07:23:20,636 [shard 0] init - Shutting down authorization cache api
INFO  2023-04-11 07:23:20,637 [shard 0] init - Shutting down authorization cache api was successful
INFO  2023-04-11 07:23:20,637 [shard 0] init - Shutting down auth service
Rate-limit: suppressed 23 backtraces on shard 0
------ Ending test debug/topology_raft_disabled.test_raft_upgrade.1::test_recovery_after_majority_loss ------
```

--- a/db/hints/manager.cc
+++ b/db/hints/manager.cc
@@ -44,7 +44,7 @@ static logging::logger manager_logger("hints_manager");
 const std::string manager::FILENAME_PREFIX("HintsLog" + commitlog::descriptor::SEPARATOR);
 
 const std::chrono::seconds manager::hint_file_write_timeout = std::chrono::seconds(2);
-const std::chrono::seconds manager::hints_flush_period = std::chrono::seconds(10);
+std::chrono::seconds manager::hints_flush_period = std::chrono::seconds(10);
 
 manager::manager(sstring hints_directory, host_filter filter, int64_t max_hint_window_ms, resource_manager& res_manager, distributed<replica::database>& db)
     : _hints_dir(fs::path(hints_directory) / format("{:d}", this_shard_id()))
@@ -52,7 +52,11 @@ manager::manager(sstring hints_directory, host_filter filter, int64_t max_hint_w
     , _max_hint_window_us(max_hint_window_ms * 1000)
     , _local_db(db.local())
     , _resource_manager(res_manager)
-{}
+{
+    if (utils::get_local_injector().enter("decrease_hints_flush_period")) {
+        hints_flush_period = std::chrono::seconds{1};
+    }
+}
 
 manager::~manager() {
     assert(_ep_managers.empty());

--- a/db/hints/manager.hh
+++ b/db/hints/manager.hh
@@ -501,7 +501,8 @@ private:
 
 public:
     static const std::string FILENAME_PREFIX;
-    static const std::chrono::seconds hints_flush_period;
+    // Non-const - can be modified with an error injection.
+    static std::chrono::seconds hints_flush_period;
     static const std::chrono::seconds hint_file_write_timeout;
 
 private:

--- a/db/view/view.cc
+++ b/db/view/view.cc
@@ -1605,7 +1605,8 @@ static future<> apply_to_remote_endpoints(service::storage_proxy& proxy, gms::in
             std::move(pending_endpoints),
             db::write_type::VIEW,
             std::move(tr_state),
-            allow_hints);
+            allow_hints,
+            service::is_cancellable::yes);
 }
 
 static bool should_update_synchronously(const schema& s) {

--- a/service/storage_proxy.cc
+++ b/service/storage_proxy.cc
@@ -1165,8 +1165,6 @@ public:
     };
 };
 
-using is_cancellable = bool_class<struct cancellable_tag>;
-
 class abstract_write_response_handler : public seastar::enable_shared_from_this<abstract_write_response_handler>, public bi::list_base_hook<bi::link_mode<bi::auto_unlink>> {
 protected:
     using error = storage_proxy::error;
@@ -2307,7 +2305,7 @@ future<result<>> storage_proxy::response_wait(storage_proxy::response_id_type id
 result<storage_proxy::response_id_type> storage_proxy::create_write_response_handler(locator::effective_replication_map_ptr ermp,
                              db::consistency_level cl, db::write_type type, std::unique_ptr<mutation_holder> m,
                              inet_address_vector_replica_set targets, const inet_address_vector_topology_change& pending_endpoints, inet_address_vector_topology_change dead_endpoints, tracing::trace_state_ptr tr_state,
-                             storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info)
+                             storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info, is_cancellable cancellable)
 {
     shared_ptr<abstract_write_response_handler> h;
     auto& rs = ermp->get_replication_strategy();
@@ -2317,7 +2315,6 @@ result<storage_proxy::response_id_type> storage_proxy::create_write_response_han
     } else if (cl == db::consistency_level::EACH_QUORUM && rs.get_type() == locator::replication_strategy_type::network_topology){
         h = ::make_shared<datacenter_sync_write_response_handler>(shared_from_this(), std::move(ermp), cl, type, std::move(m), std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info);
     } else {
-        is_cancellable cancellable{type == db::write_type::VIEW};
         h = ::make_shared<write_response_handler>(shared_from_this(), std::move(ermp), cl, type, std::move(m), std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info, cancellable);
     }
     return bo::success(register_response_handler(std::move(h)));
@@ -2843,7 +2840,7 @@ storage_proxy::mutate_counter_on_leader_and_replicate(const schema_ptr& s, froze
 
 result<storage_proxy::response_id_type>
 storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::token& token, std::unique_ptr<mutation_holder> mh,
-        db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit) {
+        db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit, is_cancellable cancellable) {
     replica::table& table = _db.local().find_column_family(s->id());
     auto erm = table.get_effective_replication_map();
     inet_address_vector_replica_set natural_endpoints = erm->get_natural_endpoints_without_node_being_replaced(token);
@@ -2907,7 +2904,7 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
     db::assure_sufficient_live_nodes(cl, *erm, live_endpoints, pending_endpoints);
 
     return create_write_response_handler(std::move(erm), cl, type, std::move(mh), std::move(live_endpoints), pending_endpoints,
-            std::move(dead_endpoints), std::move(tr_state), get_stats(), std::move(permit), rate_limit_info);
+            std::move(dead_endpoints), std::move(tr_state), get_stats(), std::move(permit), rate_limit_info, cancellable);
 }
 
 /**
@@ -2920,13 +2917,13 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
 result<storage_proxy::response_id_type>
 storage_proxy::create_write_response_handler(const mutation& m, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit) {
     return create_write_response_handler_helper(m.schema(), m.token(), std::make_unique<shared_mutation>(m), cl, type, tr_state,
-            std::move(permit), allow_limit);
+            std::move(permit), allow_limit, is_cancellable::no);
 }
 
 result<storage_proxy::response_id_type>
 storage_proxy::create_write_response_handler(const hint_wrapper& h, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit) {
     return create_write_response_handler_helper(h.mut.schema(), h.mut.token(), std::make_unique<hint_mutation>(h.mut), cl, type, tr_state,
-            std::move(permit), allow_limit);
+            std::move(permit), allow_limit, is_cancellable::yes);
 }
 
 result<storage_proxy::response_id_type>
@@ -2941,7 +2938,7 @@ storage_proxy::create_write_response_handler(const read_repair_mutation& mut, db
     tracing::trace(tr_state, "Creating write handler for read repair token: {} endpoint: {}", mh->token(), endpoints);
 
     // No rate limiting for read repair
-    return create_write_response_handler(std::move(mut.ermp), cl, type, std::move(mh), std::move(endpoints), inet_address_vector_topology_change(), inet_address_vector_topology_change(), std::move(tr_state), get_stats(), std::move(permit), std::monostate());
+    return create_write_response_handler(std::move(mut.ermp), cl, type, std::move(mh), std::move(endpoints), inet_address_vector_topology_change(), inet_address_vector_topology_change(), std::move(tr_state), get_stats(), std::move(permit), std::monostate(), is_cancellable::no);
 }
 
 result<storage_proxy::response_id_type>
@@ -2950,7 +2947,7 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
     auto& [commit, s, h, t] = meta;
 
     return create_write_response_handler_helper(s, t, std::make_unique<cas_mutation>(std::move(commit), s, std::move(h)), cl,
-            db::write_type::CAS, tr_state, std::move(permit), allow_limit);
+            db::write_type::CAS, tr_state, std::move(permit), allow_limit, is_cancellable::no);
 }
 
 result<storage_proxy::response_id_type>
@@ -2967,7 +2964,7 @@ storage_proxy::create_write_response_handler(const std::tuple<lw_shared_ptr<paxo
 
     // No rate limiting for paxos (yet)
     return create_write_response_handler(std::move(ermp), cl, db::write_type::CAS, std::make_unique<cas_mutation>(std::move(commit), s, nullptr), std::move(endpoints),
-                    inet_address_vector_topology_change(), inet_address_vector_topology_change(), std::move(tr_state), get_stats(), std::move(permit), std::monostate());
+                    inet_address_vector_topology_change(), inet_address_vector_topology_change(), std::move(tr_state), get_stats(), std::move(permit), std::monostate(), is_cancellable::no);
 }
 
 void storage_proxy::register_cdc_operation_result_tracker(const storage_proxy::unique_response_handler_vector& ids, lw_shared_ptr<cdc::operation_result_tracker> tracker) {
@@ -3495,7 +3492,7 @@ storage_proxy::mutate_atomically_result(std::vector<mutation> mutations, db::con
             return _p.mutate_prepare<>(std::array<mutation, 1>{std::move(m)}, cl, db::write_type::BATCH_LOG, _permit, [this] (const mutation& m, db::consistency_level cl, db::write_type type, service_permit permit) {
                 auto& table = _p._db.local().find_column_family(m.schema()->id());
                 auto ermp = table.get_effective_replication_map();
-                return _p.create_write_response_handler(std::move(ermp), cl, type, std::make_unique<shared_mutation>(m), _batchlog_endpoints, {}, {}, _trace_state, _stats, std::move(permit), std::monostate());
+                return _p.create_write_response_handler(std::move(ermp), cl, type, std::make_unique<shared_mutation>(m), _batchlog_endpoints, {}, {}, _trace_state, _stats, std::move(permit), std::monostate(), is_cancellable::no);
             }).then(utils::result_wrap([this, cl] (unique_response_handler_vector ids) {
                 _p.register_cdc_operation_result_tracker(ids, _cdc_tracker);
                 return _p.mutate_begin(std::move(ids), cl, _trace_state, _timeout);
@@ -3600,7 +3597,8 @@ future<> storage_proxy::send_to_endpoint(
         db::write_type type,
         tracing::trace_state_ptr tr_state,
         write_stats& stats,
-        allow_hints allow_hints) {
+        allow_hints allow_hints,
+        is_cancellable cancellable) {
     utils::latency_counter lc;
     lc.start();
 
@@ -3612,7 +3610,7 @@ future<> storage_proxy::send_to_endpoint(
         timeout = clock_type::now() + 5min;
     }
     return mutate_prepare(std::array{std::move(m)}, cl, type, /* does view building should hold a real permit */ empty_service_permit(),
-            [this, tr_state, target = std::array{target}, pending_endpoints = std::move(pending_endpoints), &stats] (
+            [this, tr_state, target = std::array{target}, pending_endpoints = std::move(pending_endpoints), &stats, cancellable] (
                 std::unique_ptr<mutation_holder>& m,
                 db::consistency_level cl,
                 db::write_type type, service_permit permit) mutable {
@@ -3639,7 +3637,8 @@ future<> storage_proxy::send_to_endpoint(
             tr_state,
             stats,
             std::move(permit),
-            std::monostate()); // TODO: Pass the correct enforcement type
+            std::monostate(), // TODO: Pass the correct enforcement type
+            cancellable);
     }).then(utils::result_wrap([this, cl, tr_state = std::move(tr_state), timeout = std::move(timeout)] (unique_response_handler_vector ids) mutable {
         return mutate_begin(std::move(ids), cl, std::move(tr_state), std::move(timeout));
     })).then_wrapped([p = shared_from_this(), lc, &stats] (future<result<>> f) {
@@ -3653,7 +3652,8 @@ future<> storage_proxy::send_to_endpoint(
         inet_address_vector_topology_change pending_endpoints,
         db::write_type type,
         tracing::trace_state_ptr tr_state,
-        allow_hints allow_hints) {
+        allow_hints allow_hints,
+        is_cancellable cancellable) {
     return send_to_endpoint(
             std::make_unique<shared_mutation>(std::move(fm_a_s)),
             std::move(target),
@@ -3661,7 +3661,8 @@ future<> storage_proxy::send_to_endpoint(
             type,
             std::move(tr_state),
             get_stats(),
-            allow_hints);
+            allow_hints,
+            cancellable);
 }
 
 future<> storage_proxy::send_to_endpoint(
@@ -3671,7 +3672,8 @@ future<> storage_proxy::send_to_endpoint(
         db::write_type type,
         tracing::trace_state_ptr tr_state,
         write_stats& stats,
-        allow_hints allow_hints) {
+        allow_hints allow_hints,
+        is_cancellable cancellable) {
     return send_to_endpoint(
             std::make_unique<shared_mutation>(std::move(fm_a_s)),
             std::move(target),
@@ -3679,7 +3681,8 @@ future<> storage_proxy::send_to_endpoint(
             type,
             std::move(tr_state),
             stats,
-            allow_hints);
+            allow_hints,
+            cancellable);
 }
 
 future<> storage_proxy::send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s, gms::inet_address target) {
@@ -3691,7 +3694,8 @@ future<> storage_proxy::send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s,
                 db::write_type::SIMPLE,
                 tracing::trace_state_ptr(),
                 get_stats(),
-                allow_hints::no);
+                allow_hints::no,
+                is_cancellable::yes);
     }
 
     return send_to_endpoint(
@@ -3701,7 +3705,8 @@ future<> storage_proxy::send_hint_to_endpoint(frozen_mutation_and_schema fm_a_s,
             db::write_type::SIMPLE,
             tracing::trace_state_ptr(),
             get_stats(),
-            allow_hints::no);
+            allow_hints::no,
+            is_cancellable::yes);
 }
 
 future<> storage_proxy::send_hint_to_all_replicas(frozen_mutation_and_schema fm_a_s) {

--- a/service/storage_proxy.cc
+++ b/service/storage_proxy.cc
@@ -1165,7 +1165,9 @@ public:
     };
 };
 
-class abstract_write_response_handler : public seastar::enable_shared_from_this<abstract_write_response_handler> {
+using is_cancellable = bool_class<struct cancellable_tag>;
+
+class abstract_write_response_handler : public seastar::enable_shared_from_this<abstract_write_response_handler>, public bi::list_base_hook<bi::link_mode<bi::auto_unlink>> {
 protected:
     using error = storage_proxy::error;
     storage_proxy::response_id_type _id;
@@ -1211,7 +1213,7 @@ public:
             db::consistency_level cl, db::write_type type,
             std::unique_ptr<mutation_holder> mh, inet_address_vector_replica_set targets, tracing::trace_state_ptr trace_state,
             storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info, size_t pending_endpoints = 0,
-            inet_address_vector_topology_change dead_endpoints = {})
+            inet_address_vector_topology_change dead_endpoints = {}, is_cancellable cancellable = is_cancellable::no)
             : _id(p->get_next_response_id()), _proxy(std::move(p))
             , _effective_replication_map_ptr(std::move(erm))
             , _trace_state(trace_state), _cl(cl), _type(type), _mutation_holder(std::move(mh)), _targets(std::move(targets)),
@@ -1222,6 +1224,10 @@ public:
         // or we may fail the consistency level guarantees (see #833, #8058)
         _total_block_for = db::block_for(*_effective_replication_map_ptr, _cl) + pending_endpoints;
         ++_stats.writes;
+
+        if (cancellable) {
+            register_cancellable();
+        }
     }
     virtual ~abstract_write_response_handler() {
         --_stats.writes;
@@ -1249,6 +1255,8 @@ public:
                 _cdc_operation_result_tracker->on_mutation_failed();
             }
         }
+
+        update_cancellable_live_iterators();
     }
     bool is_counter() const {
         return _type == db::write_type::COUNTER;
@@ -1456,6 +1464,11 @@ public:
         return _stats;
     }
     friend storage_proxy;
+
+private:
+    void register_cancellable();
+    // Called on destruction
+    void update_cancellable_live_iterators();
 };
 
 class datacenter_write_response_handler : public abstract_write_response_handler {
@@ -1488,33 +1501,16 @@ public:
             db::consistency_level cl, db::write_type type,
             std::unique_ptr<mutation_holder> mh, inet_address_vector_replica_set targets,
             const inet_address_vector_topology_change& pending_endpoints, inet_address_vector_topology_change dead_endpoints, tracing::trace_state_ptr tr_state,
-            storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info) :
+            storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info, is_cancellable cancellable) :
                 abstract_write_response_handler(std::move(p), std::move(ermp), cl, type, std::move(mh),
-                        std::move(targets), std::move(tr_state), stats, std::move(permit), rate_limit_info, pending_endpoints.size(), std::move(dead_endpoints)) {
+                        std::move(targets), std::move(tr_state), stats, std::move(permit), rate_limit_info, pending_endpoints.size(), std::move(dead_endpoints), cancellable) {
         _total_endpoints = _targets.size();
     }
 };
 
-class view_update_write_response_handler : public write_response_handler, public bi::list_base_hook<bi::link_mode<bi::auto_unlink>> {
-public:
-    view_update_write_response_handler(shared_ptr<storage_proxy> p,
-            locator::effective_replication_map_ptr ermp,
-            db::consistency_level cl,
-            std::unique_ptr<mutation_holder> mh, inet_address_vector_replica_set targets,
-            const inet_address_vector_topology_change& pending_endpoints, inet_address_vector_topology_change dead_endpoints, tracing::trace_state_ptr tr_state,
-            storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info):
-                write_response_handler(p, std::move(ermp), cl, db::write_type::VIEW, std::move(mh),
-                        std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info) {
-        register_in_intrusive_list(*p);
-    }
-    ~view_update_write_response_handler();
-private:
-    void register_in_intrusive_list(storage_proxy& p);
-};
-
-class storage_proxy::view_update_handlers_list : public bi::list<view_update_write_response_handler, bi::base_hook<view_update_write_response_handler>, bi::constant_time_size<false>> {
+class storage_proxy::view_update_handlers_list : public bi::list<abstract_write_response_handler, bi::base_hook<abstract_write_response_handler>, bi::constant_time_size<false>> {
     // _live_iterators holds all iterators that point into the bi:list in the base class of this object.
-    // If we remove a view_update_write_response_handler from the list, and an iterator happens to point
+    // If we remove a abstract_write_response_handler from the list, and an iterator happens to point
     // into it, we advance the iterator so it doesn't point at a removed object. See #4912.
     std::vector<iterator*> _live_iterators;
 public:
@@ -1527,7 +1523,7 @@ public:
     void unregister_live_iterator(iterator* itp) {
         _live_iterators.erase(boost::remove(_live_iterators, itp), _live_iterators.end());
     }
-    void update_live_iterators(view_update_write_response_handler* vuwrh) {
+    void update_live_iterators(abstract_write_response_handler* vuwrh) {
         // vuwrh is being removed from the b::list, so if any live iterator points at it,
         // move it to the next object (this requires that the list is traversed in the forward
         // direction).
@@ -1550,13 +1546,15 @@ public:
     };
 };
 
-void view_update_write_response_handler::register_in_intrusive_list(storage_proxy& p) {
-    p.get_view_update_handlers_list().push_back(*this);
+void abstract_write_response_handler::register_cancellable() {
+    _proxy->_view_update_handlers_list->push_back(*this);
 }
 
 
-view_update_write_response_handler::~view_update_write_response_handler() {
-    _proxy->_view_update_handlers_list->update_live_iterators(this);
+void abstract_write_response_handler::update_cancellable_live_iterators() {
+    if (is_linked()) {
+        _proxy->_view_update_handlers_list->update_live_iterators(this);
+    }
 }
 
 class datacenter_sync_write_response_handler : public abstract_write_response_handler {
@@ -2313,10 +2311,9 @@ result<storage_proxy::response_id_type> storage_proxy::create_write_response_han
         h = ::make_shared<datacenter_write_response_handler>(shared_from_this(), std::move(ermp), cl, type, std::move(m), std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info);
     } else if (cl == db::consistency_level::EACH_QUORUM && rs.get_type() == locator::replication_strategy_type::network_topology){
         h = ::make_shared<datacenter_sync_write_response_handler>(shared_from_this(), std::move(ermp), cl, type, std::move(m), std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info);
-    } else if (type == db::write_type::VIEW) {
-        h = ::make_shared<view_update_write_response_handler>(shared_from_this(), std::move(ermp), cl, std::move(m), std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info);
     } else {
-        h = ::make_shared<write_response_handler>(shared_from_this(), std::move(ermp), cl, type, std::move(m), std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info);
+        is_cancellable cancellable{type == db::write_type::VIEW};
+        h = ::make_shared<write_response_handler>(shared_from_this(), std::move(ermp), cl, type, std::move(m), std::move(targets), pending_endpoints, std::move(dead_endpoints), std::move(tr_state), stats, std::move(permit), rate_limit_info, cancellable);
     }
     return bo::success(register_response_handler(std::move(h)));
 }

--- a/service/storage_proxy.cc
+++ b/service/storage_proxy.cc
@@ -1508,13 +1508,18 @@ public:
     }
 };
 
-class storage_proxy::view_update_handlers_list : public bi::list<abstract_write_response_handler, bi::base_hook<abstract_write_response_handler>, bi::constant_time_size<false>> {
+// This list contains `abstract_write_response_handler`s which were constructed as `cancellable`.
+// When a `cancellable` handler is constructed, it adds itself to the list (see `register_cancellable`).
+// We use the list to cancel handlers - as if the write timed out - on certain events, such as when
+// we shutdown a node so that shutdown is not blocked.
+// We don't add normal data path writes to the list, only background work such as hints and view updates.
+class storage_proxy::cancellable_write_handlers_list : public bi::list<abstract_write_response_handler, bi::base_hook<abstract_write_response_handler>, bi::constant_time_size<false>> {
     // _live_iterators holds all iterators that point into the bi:list in the base class of this object.
     // If we remove a abstract_write_response_handler from the list, and an iterator happens to point
     // into it, we advance the iterator so it doesn't point at a removed object. See #4912.
     std::vector<iterator*> _live_iterators;
 public:
-    view_update_handlers_list() {
+    cancellable_write_handlers_list() {
         _live_iterators.reserve(10); // We only expect 1.
     }
     void register_live_iterator(iterator* itp) noexcept { // We don't tolerate failure, so abort instead
@@ -1523,37 +1528,37 @@ public:
     void unregister_live_iterator(iterator* itp) {
         _live_iterators.erase(boost::remove(_live_iterators, itp), _live_iterators.end());
     }
-    void update_live_iterators(abstract_write_response_handler* vuwrh) {
-        // vuwrh is being removed from the b::list, so if any live iterator points at it,
+    void update_live_iterators(abstract_write_response_handler* handler) {
+        // handler is being removed from the b::list, so if any live iterator points at it,
         // move it to the next object (this requires that the list is traversed in the forward
         // direction).
         for (auto& itp : _live_iterators) {
-            if (&**itp == vuwrh) {
+            if (&**itp == handler) {
                 ++*itp;
             }
         }
     }
     class iterator_guard {
-        view_update_handlers_list& _vuhl;
+        cancellable_write_handlers_list& _handlers;
         iterator* _itp;
     public:
-        iterator_guard(view_update_handlers_list& vuhl, iterator& it) : _vuhl(vuhl), _itp(&it) {
-            _vuhl.register_live_iterator(_itp);
+        iterator_guard(cancellable_write_handlers_list& handlers, iterator& it) : _handlers(handlers), _itp(&it) {
+            _handlers.register_live_iterator(_itp);
         }
         ~iterator_guard() {
-            _vuhl.unregister_live_iterator(_itp);
+            _handlers.unregister_live_iterator(_itp);
         }
     };
 };
 
 void abstract_write_response_handler::register_cancellable() {
-    _proxy->_view_update_handlers_list->push_back(*this);
+    _proxy->_cancellable_write_handlers_list->push_back(*this);
 }
 
 
 void abstract_write_response_handler::update_cancellable_live_iterators() {
     if (is_linked()) {
-        _proxy->_view_update_handlers_list->update_live_iterators(this);
+        _proxy->_cancellable_write_handlers_list->update_live_iterators(this);
     }
 }
 
@@ -2721,7 +2726,7 @@ storage_proxy::storage_proxy(distributed<replica::database>& db, gms::gossiper& 
     , _background_write_throttle_threahsold(cfg.available_memory / 10)
     , _mutate_stage{"storage_proxy_mutate", &storage_proxy::do_mutate}
     , _max_view_update_backlog(max_view_update_backlog)
-    , _view_update_handlers_list(std::make_unique<view_update_handlers_list>()) {
+    , _cancellable_write_handlers_list(std::make_unique<cancellable_write_handlers_list>()) {
     namespace sm = seastar::metrics;
     _metrics.add_group(storage_proxy_stats::COORDINATOR_STATS_CATEGORY, {
         sm::make_queue_length("current_throttled_writes", [this] { return _throttled_writes.size(); },
@@ -6236,24 +6241,24 @@ void storage_proxy::on_leave_cluster(const gms::inet_address& endpoint) {
 
 void storage_proxy::on_up(const gms::inet_address& endpoint) {};
 
-void storage_proxy::retire_view_response_handlers(noncopyable_function<bool(const abstract_write_response_handler&)> filter_fun) {
+void storage_proxy::cancel_write_handlers(noncopyable_function<bool(const abstract_write_response_handler&)> filter_fun) {
     assert(thread::running_in_thread());
-    auto it = _view_update_handlers_list->begin();
-    while (it != _view_update_handlers_list->end()) {
+    auto it = _cancellable_write_handlers_list->begin();
+    while (it != _cancellable_write_handlers_list->end()) {
         auto guard = it->shared_from_this();
         if (filter_fun(*it) && _response_handlers.contains(it->id())) {
             it->timeout_cb();
         }
         ++it;
         if (need_preempt()) {
-            view_update_handlers_list::iterator_guard ig{*_view_update_handlers_list, it};
+            cancellable_write_handlers_list::iterator_guard ig{*_cancellable_write_handlers_list, it};
             seastar::thread::yield();
         }
     }
 }
 
 void storage_proxy::on_down(const gms::inet_address& endpoint) {
-    return retire_view_response_handlers([endpoint] (const abstract_write_response_handler& handler) {
+    return cancel_write_handlers([endpoint] (const abstract_write_response_handler& handler) {
         const auto& targets = handler.get_targets();
         return boost::find(targets, endpoint) != targets.end();
     });
@@ -6263,7 +6268,7 @@ future<> storage_proxy::drain_on_shutdown() {
     //NOTE: the thread is spawned here because there are delicate lifetime issues to consider
     // and writing them down with plain futures is error-prone.
     return async([this] {
-        retire_view_response_handlers([] (const abstract_write_response_handler&) { return true; });
+        cancel_write_handlers([] (const abstract_write_response_handler&) { return true; });
         _hints_resource_manager.stop().get();
     });
 }

--- a/service/storage_proxy.hh
+++ b/service/storage_proxy.hh
@@ -470,10 +470,6 @@ public:
         return _cdc;
     }
 
-    view_update_handlers_list& get_view_update_handlers_list() {
-        return *_view_update_handlers_list;
-    }
-
     response_id_type get_next_response_id() {
         auto next = _next_response_id++;
         if (next == 0) { // 0 is reserved for unique_response_handler

--- a/service/storage_proxy.hh
+++ b/service/storage_proxy.hh
@@ -99,6 +99,8 @@ struct view_update_backlog_timestamped {
 struct allow_hints_tag {};
 using allow_hints = bool_class<allow_hints_tag>;
 
+using is_cancellable = bool_class<struct cancellable_tag>;
+
 struct storage_proxy_coordinator_query_result {
     foreign_ptr<lw_shared_ptr<query::result>> query_result;
     replicas_per_token_range last_replicas;
@@ -307,9 +309,9 @@ private:
     ::shared_ptr<abstract_write_response_handler>& get_write_response_handler(storage_proxy::response_id_type id);
     result<response_id_type> create_write_response_handler_helper(schema_ptr s, const dht::token& token,
             std::unique_ptr<mutation_holder> mh, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state,
-            service_permit permit, db::allow_per_partition_rate_limit allow_limit);
+            service_permit permit, db::allow_per_partition_rate_limit allow_limit, is_cancellable);
     result<response_id_type> create_write_response_handler(locator::effective_replication_map_ptr ermp, db::consistency_level cl, db::write_type type, std::unique_ptr<mutation_holder> m, inet_address_vector_replica_set targets,
-            const inet_address_vector_topology_change& pending_endpoints, inet_address_vector_topology_change, tracing::trace_state_ptr tr_state, storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info);
+            const inet_address_vector_topology_change& pending_endpoints, inet_address_vector_topology_change, tracing::trace_state_ptr tr_state, storage_proxy::write_stats& stats, service_permit permit, db::per_partition_rate_limit::info rate_limit_info, is_cancellable);
     result<response_id_type> create_write_response_handler(const mutation&, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit);
     result<response_id_type> create_write_response_handler(const hint_wrapper&, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit);
     result<response_id_type> create_write_response_handler(const read_repair_mutation&, db::consistency_level cl, db::write_type type, tracing::trace_state_ptr tr_state, service_permit permit, db::allow_per_partition_rate_limit allow_limit);
@@ -416,7 +418,8 @@ private:
             db::write_type type,
             tracing::trace_state_ptr tr_state,
             write_stats& stats,
-            allow_hints allow_hints = allow_hints::yes);
+            allow_hints,
+            is_cancellable);
 
     db::view::update_backlog get_view_update_backlog() const;
 
@@ -565,9 +568,9 @@ public:
     // hinted handoff support, and just one target. See also
     // send_to_live_endpoints() - another take on the same original function.
     future<> send_to_endpoint(frozen_mutation_and_schema fm_a_s, gms::inet_address target, inet_address_vector_topology_change pending_endpoints, db::write_type type,
-            tracing::trace_state_ptr tr_state, write_stats& stats, allow_hints allow_hints = allow_hints::yes);
+            tracing::trace_state_ptr tr_state, write_stats& stats, allow_hints, is_cancellable);
     future<> send_to_endpoint(frozen_mutation_and_schema fm_a_s, gms::inet_address target, inet_address_vector_topology_change pending_endpoints, db::write_type type,
-            tracing::trace_state_ptr tr_state, allow_hints allow_hints = allow_hints::yes);
+            tracing::trace_state_ptr tr_state, allow_hints, is_cancellable);
 
     // Send a mutation to a specific remote target as a hint.
     // Unlike regular mutations during write operations, hints are sent on the streaming connection

--- a/service/storage_proxy.hh
+++ b/service/storage_proxy.hh
@@ -273,8 +273,8 @@ private:
     std::unordered_map<gms::inet_address, view_update_backlog_timestamped> _view_update_backlogs;
 
     //NOTICE(sarna): This opaque pointer is here just to avoid moving write handler class definitions from .cc to .hh. It's slow path.
-    class view_update_handlers_list;
-    std::unique_ptr<view_update_handlers_list> _view_update_handlers_list;
+    class cancellable_write_handlers_list;
+    std::unique_ptr<cancellable_write_handlers_list> _cancellable_write_handlers_list;
 
     /* This is a pointer to the shard-local part of the sharded cdc_service:
      * storage_proxy needs access to cdc_service to augument mutations.
@@ -427,7 +427,8 @@ private:
     template<typename Range>
     future<> mutate_counters(Range&& mutations, db::consistency_level cl, tracing::trace_state_ptr tr_state, service_permit permit, clock_type::time_point timeout);
 
-    void retire_view_response_handlers(noncopyable_function<bool(const abstract_write_response_handler&)> filter_fun);
+    // Retires (times out) write response handlers which were constructed as `cancellable` and pass the given filter.
+    void cancel_write_handlers(noncopyable_function<bool(const abstract_write_response_handler&)> filter_fun);
 
     /**
      * Returns whether for a range query doing a query against merged is likely

--- a/test/pylib/manager_client.py
+++ b/test/pylib/manager_client.py
@@ -161,7 +161,7 @@ class ManagerClient():
         await self.server_sees_others(server_id, wait_others, interval = wait_interval)
         self._driver_update()
 
-    async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None, cmdline: Optional[List[str]] = None, config: Optional[dict[str, str]] = None, start: bool = True) -> ServerInfo:
+    async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None, cmdline: Optional[List[str]] = None, config: Optional[dict[str, Any]] = None, start: bool = True) -> ServerInfo:
         """Add a new server"""
         try:
             data: dict[str, Any] = {'start': start}

--- a/test/pylib/scylla_cluster.py
+++ b/test/pylib/scylla_cluster.py
@@ -18,7 +18,7 @@ import shutil
 import tempfile
 import time
 import traceback
-from typing import Optional, Dict, List, Set, Tuple, Callable, AsyncIterator, NamedTuple, Union
+from typing import Any, Optional, Dict, List, Set, Tuple, Callable, AsyncIterator, NamedTuple, Union
 import uuid
 from enum import Enum
 from io import BufferedWriter
@@ -203,7 +203,7 @@ class ScyllaServer:
                  logger: Union[logging.Logger, logging.LoggerAdapter],
                  cluster_name: str, ip_addr: str, seeds: List[str],
                  cmdline_options: List[str],
-                 config_options: Dict[str, str]) -> None:
+                 config_options: Dict[str, Any]) -> None:
         # pylint: disable=too-many-arguments
         self.server_id = ServerNum(ScyllaServer.newid())
         self.exe = pathlib.Path(exe).resolve()
@@ -565,7 +565,7 @@ class ScyllaCluster:
         cluster_name: str
         ip_addr: IPAddress
         seeds: List[str]
-        config_from_test: dict[str, str]
+        config_from_test: dict[str, Any]
         cmdline_from_test: List[str]
 
     def __init__(self, logger: Union[logging.Logger, logging.LoggerAdapter],
@@ -652,11 +652,11 @@ class ScyllaCluster:
     def _seeds(self) -> List[str]:
         return [server.ip_addr for server in self.running.values()]
 
-    async def add_server(self, replace_cfg: Optional[ReplaceConfig] = None, cmdline: Optional[List[str]] = None, config: Optional[dict[str, str]] = None, start: bool = True) -> ServerInfo:
+    async def add_server(self, replace_cfg: Optional[ReplaceConfig] = None, cmdline: Optional[List[str]] = None, config: Optional[dict[str, Any]] = None, start: bool = True) -> ServerInfo:
         """Add a new server to the cluster"""
         self.is_dirty = True
 
-        extra_config: dict[str, str] = config.copy() if config else {}
+        extra_config: dict[str, Any] = config.copy() if config else {}
         if replace_cfg:
             replaced_id = replace_cfg.replaced_id
             assert replaced_id in self.servers, \

--- a/test/pylib/scylla_cluster.py
+++ b/test/pylib/scylla_cluster.py
@@ -502,9 +502,15 @@ class ScyllaServer:
         except ProcessLookupError:
             pass
         else:
-            # FIXME: add timeout, fail the test and mark cluster as dirty
-            # if we timeout.
-            await self.cmd.wait()
+            STOP_TIMEOUT_SECONDS = 60
+            wait_task = self.cmd.wait()
+            try:
+                await asyncio.wait_for(wait_task, timeout=STOP_TIMEOUT_SECONDS)
+            except asyncio.TimeoutError:
+                self.cmd.kill()
+                await self.cmd.wait()
+                raise RuntimeError(
+                    f"Stopping server {self} gracefully took longer than {STOP_TIMEOUT_SECONDS}s")
         finally:
             if self.cmd:
                 self.logger.info("gracefully stopped %s", self)

--- a/test/topology_custom/suite.yaml
+++ b/test/topology_custom/suite.yaml
@@ -5,3 +5,7 @@ cluster:
 extra_scylla_config_options:
     authenticator: AllowAllAuthenticator
     authorizer: AllowAllAuthorizer
+skip_in_release:
+  - test_shutdown_hang
+skip_in_debug:
+  - test_shutdown_hang

--- a/test/topology_custom/test_shutdown_hang.py
+++ b/test/topology_custom/test_shutdown_hang.py
@@ -1,0 +1,63 @@
+import asyncio
+import logging
+import time
+import pytest
+
+from cassandra.query import SimpleStatement # type: ignore
+from cassandra.cluster import ConsistencyLevel # type: ignore
+from cassandra.protocol import WriteTimeout # type: ignore
+
+from test.pylib.manager_client import ManagerClient
+from test.topology.util import wait_for_token_ring_and_group0_consistency
+
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.asyncio
+async def test_hints_manager_shutdown_hang(manager: ManagerClient) -> None:
+    """Reproducer for #8079"""
+    s1 = await manager.server_add(config={
+        'error_injections_at_startup': ['decrease_hints_flush_period']
+    })
+    s2 = await manager.server_add()
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
+
+    cql = manager.get_cql()
+
+    logger.info("Create keyspace and table")
+    await cql.run_async("create keyspace ks with replication = {'class': 'SimpleStrategy', 'replication_factor': 2}")
+    await cql.run_async("create table ks.t (pk int primary key)")
+
+    logger.info(f"Stop {s2}")
+    await manager.server_stop(s2.server_id)
+
+    logger.info("Write data with small timeout")
+    # We're using a small timeout for the insert so it's not unexpected that it would fail on slow
+    # CI machines. To avoid flakiness we disable the test in debug mode (as well as release since
+    # it requires an error injection - so it will run only in dev mode) and we retry the write 10 times.
+    passed = False
+    for _ in range(10):
+        try:
+            await cql.run_async(SimpleStatement("insert into ks.t (pk) values (0) using timeout 500ms",
+                                                consistency_level=ConsistencyLevel.ONE))
+        except WriteTimeout:
+            logger.info("write timeout, retrying")
+        else:
+            passed = True
+            break
+
+    if not passed:
+        pytest.fail("Write timed out on each attempt")
+
+    # The write succeeded but a background task was left to finish the write to the other node
+    # (which is dead but the first node didn't mark it as dead yet).
+    # The background task will timeout shortly because of 'using timeout' in the statement.
+    # This will cause a hint to get created.
+    # The hints manager starts sending the hint soon after (hint flushing happens every
+    # ~1 second with the error injection).
+    logger.info("Sleep")
+    await asyncio.sleep(2)
+
+    logger.info(f"Stop {s1} gracefully")
+    await manager.server_stop_gracefully(s1.server_id)

